micrometer-core/src/main/java/io/micrometer/core/instrument/binder/neo4j/Neo4jMetrics.java

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright 2020 VMware, Inc.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.neo4j;
+
+import io.micrometer.core.instrument.FunctionCounter;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.instrument.binder.MeterBinder;
+import io.micrometer.core.lang.NonNullApi;
+import io.micrometer.core.lang.NonNullFields;
+
+import java.util.function.Consumer;
+
+import org.neo4j.driver.ConnectionPoolMetrics;
+import org.neo4j.driver.Driver;
+import org.neo4j.driver.Metrics;
+
+/**
+ * {@link MeterBinder} that binds all available Neo4j driver metrics.
+ *
+ * @author Michael J. Simons
+ * @author Gerrit Meier
+ */
+@NonNullApi
+@NonNullFields
+public class Neo4jMetrics implements MeterBinder {
+
+    /**
+     * Prefixed used for driver metrics.
+     */
+    public static final String PREFIX = "neo4j.driver.connections";
+
+    private static final String BASE_UNIT_CONNECTIONS = "connections";
+
+    private final Driver driver;
+
+    private final Iterable<Tag> tags;
+
+    public Neo4jMetrics(String name, Driver driver, Iterable<Tag> tags) {
+        this.driver = driver;
+        this.tags = Tags.concat(tags, "name", name);
+    }
+
+    @Override
+    public void bindTo(MeterRegistry meterRegistry) {
+        if (!this.driver.isMetricsEnabled()) {
+            return;
+        }
+        Metrics metrics = this.driver.metrics();
+        metrics.connectionPoolMetrics().forEach(this.getPoolMetricsBinder(meterRegistry));
```

Comment:
I was reading the JavaDoc for `connectionPoolMetrics()` and it mentions that connection pools are dynamically added and removed. It seems like we will only get metrics for whatever pools are initially available at binding time. Won't this be an issue when a pool is removed or added?

Reply:
This _could_ be the case but on the other hand it could also be a collection that does not change at runtime in practice.
I am waiting for our driver team to explain it to me ;) Will return if I get any insights.

Reply:
Thank you for checking on it. This could be important enough to reconsider including the metrics as is in our upcoming release.

If pools do dynamically change, we might need to periodically check, or better yet, have some callback for when it changes. Or even better yet, move the micrometer instrumentation to the driver where it should be easier to handle pool changes than from outside the driver.

Reply:
I talked with the team and they explained that the `connectionPoolMetrics` could change if the topology changes in the cluster. Since we transferred this bit in this form from Spring Boot, I would say: let's keep it as it is right now but I can foresee a change for more dynamic pool support incoming. I will investigate more in the upcoming weeks.

Reply:
> Since we transferred this bit in this form from Spring Boot

I might be missing some context, but I assume this is referencing the previous pull request to Spring Boot, which didn't accept the metrics changes into Spring Boot. So there is no prior released version of this that users are already relying on, if I'm not mistaken. I would rather not release a version we already know will have issues for users when topology changes happen.

We can get better metrics with a Micrometer-based implementation of `ConnectionPoolMetrics` in the neo4j-java-driver. Is that something the driver team would consider with an optional Micrometer dependency? We could look at contributing that over there instead of this here. We could then get proper Timer metrics on the connection pools also since we would have access to individual timing recordings, which we can't do with the current approach here.